liquidSVM

PhilippPro opened the pull request:
This is the implementation of liquidSVM as mlr-learner. The probability predictions are still technically flawed and they are not (yet) responding to my emails, so unfortunately probability predictions are not provided at the moment.

larskotthoff commented:
Could you add some tests please?

PhilippPro commented:
Why are there some errors on the pr and not on the push? I also do not understand the errors very well...

larskotthoff commented:
Ok, it looks like the tests are failing with a segmentation fault. I've tried on two different machines and can't reproduce it. Does it work when you run the tests locally?

PhilippPro commented:
Yes, they were ok. The algorithm only produces warnings in the default version which is not good. Also the probability prediction is not working properly yet. But the package has quite potential regarding performance and runtime and I already contacted the creators.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+    package = "liquidSVM",
+    par.set = makeParamSet(
+      makeIntegerLearnerParam(id = "d", lower = 0L),
+      makeLogicalLearnerParam(id = "scale"),
```
Default for this is TRUE.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+    cl = "classif.liquidSVM",
+    package = "liquidSVM",
+    par.set = makeParamSet(
+      makeIntegerLearnerParam(id = "d", lower = 0L),
```
This should not be tunable (`tunable = FALSE`).

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+    par.set = makeParamSet(
+      makeIntegerLearnerParam(id = "d", lower = 0L),
+      makeLogicalLearnerParam(id = "scale"),
+      makeIntegerLearnerParam(id = "threads", lower = -1L),
```
Default is 0.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeIntegerLearnerParam(id = "d", lower = 0L),
+      makeLogicalLearnerParam(id = "scale"),
+      makeIntegerLearnerParam(id = "threads", lower = -1L),
+      makeIntegerLearnerParam(id = "partition_choice", lower = 0L),
```
Default is 0.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeLogicalLearnerParam(id = "scale"),
+      makeIntegerLearnerParam(id = "threads", lower = -1L),
+      makeIntegerLearnerParam(id = "partition_choice", lower = 0L),
+      makeIntegerLearnerParam(id = "grid_choice", lower = -2L, upper = 2L),
```
How does this interact with `gamma_steps`, `lambda_steps`, etc? Are these options mutually exclusive?

larskotthoff replied:
And all of these options should be not tunable.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeIntegerLearnerParam(id = "threads", lower = -1L),
+      makeIntegerLearnerParam(id = "partition_choice", lower = 0L),
+      makeIntegerLearnerParam(id = "grid_choice", lower = -2L, upper = 2L),
+      makeIntegerLearnerParam(id = "adaptivity_control", lower = 0L, upper = 2L),
```
Default is 0.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeIntegerLearnerParam(id = "partition_choice", lower = 0L),
+      makeIntegerLearnerParam(id = "grid_choice", lower = -2L, upper = 2L),
+      makeIntegerLearnerParam(id = "adaptivity_control", lower = 0L, upper = 2L),
+      makeIntegerLearnerParam(id = "random_seed", default = 1),
```
I didn't find any documentation on what the default random seed is -- where did you find this? It also shouldn't be tunable.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeIntegerLearnerParam(id = "grid_choice", lower = -2L, upper = 2L),
+      makeIntegerLearnerParam(id = "adaptivity_control", lower = 0L, upper = 2L),
+      makeIntegerLearnerParam(id = "random_seed", default = 1),
+      makeIntegerLearnerParam(id = "fold"),
```
Should that be `folds`? And there should be a minimum for this (and again not tunable).

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeIntegerLearnerParam(id = "adaptivity_control", lower = 0L, upper = 2L),
+      makeIntegerLearnerParam(id = "random_seed", default = 1),
+      makeIntegerLearnerParam(id = "fold"),
+      makeIntegerLearnerParam(id = "clipping"),
```
Seems that the minimum is -1. Also probably shouldn't be tunable.

larskotthoff commented on R/RLearner_classif_liquidSVM.R:
```diff
@@ -0,0 +1,49 @@
+      makeNumericLearnerParam(id = "max_lambda", lower = 0),
+      makeNumericVectorLearnerParam(id = "lambdas", lower = 0),
+      makeNumericVectorLearnerParam(id = "c_values", lower = 0),
+      makeLogicalLearnerParam(id = "useCells")
```
Seems that the default for this is `FALSE`. Should be not tunable.

larskotthoff commented on R/RLearner_regr_liquidSVM.R:
```diff
@@ -0,0 +1,43 @@
+  makeRLearnerRegr(
+    cl = "regr.liquidSVM",
+    package = "liquidSVM",
+    par.set = makeParamSet(
```
See comments for `classif` param set.

PhilippPro commented:
I tried to change all parameters according to your suggestions. I am unsure about clipping, if it is tunable or not, that's why I did not set `tunable = FALSE`.

larskotthoff commented:
Thanks, looks good -- what about restrictions between parameters that control the tuning though? If you set all the gamma parameters at the same time, I guess that some of them override others? It may make sense to add requirements saying that if you set a certain parameter, others cannot be set to avoid surprising results when users accidentally set conflicting parameters.

PhilippPro commented:
> Thanks, looks good -- what about restrictions between parameters that control the tuning though? If you set all the gamma parameters at the same time, I guess that some of them override others? It may make sense to add requirements saying that if you set a certain parameter, others cannot be set to avoid surprising results when users accidentally set conflicting parameters.

I changed the parameter settings to take into account requirements. Moreover I looked into https://github.com/cran/liquidSVM/blob/master/R/mlr.R and changed the parameter settings slightly.

larskotthoff commented:
Thanks for all your work on this. Merging.
